presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java

```diff
@@ -0,0 +1,265 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.tracer;
+
+import io.prestosql.spi.eventlistener.TracerEvent;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+// Tracer keeps track of Presto execution unit contexts and sends debugging purpose events via its emitter.
+// Users should not build complex metrics based on the action events.
+public class DefaultTracer
+        implements Tracer
+{
+    private final Consumer<TracerEvent> emitter;
+    private final Function<Map<String, Object>, String> payloadEncoder;
+    // presto coordinator or worker node emitting tracer events
+    private final String nodeId;
+    private final URI uri;
+    private final String queryId;
+    private final Optional<String> stageId;
+    private final Optional<String> taskId;
+    private final Optional<String> pipelineId;
+    private final Optional<String> splitId;
+    private final Optional<String> operatorId;
+
+    private final boolean enabled;
+
+    private DefaultTracer(
+            Consumer<TracerEvent> emitter,
+            Function<Map<String, Object>, String> payloadEncoder,
+            String nodeId,
+            URI uri,
+            String queryId,
+            Optional<String> stageId,
+            Optional<String> taskId,
+            Optional<String> pipelineId,
+            Optional<String> splitId,
+            Optional<String> operatorId,
+            boolean isEnabled)
+    {
+        this.emitter = requireNonNull(emitter, "emitter is null");
+        this.payloadEncoder = requireNonNull(payloadEncoder, "payloadEncoder is null");
+        this.nodeId = requireNonNull(nodeId, "nodeId is null");
+        this.uri = requireNonNull(uri, "uri is null");
+        this.queryId = requireNonNull(queryId, "queryId is null");
+        this.stageId = requireNonNull(stageId, "stageId is null");
+        this.taskId = requireNonNull(taskId, "taskId is null");
+        this.pipelineId = requireNonNull(pipelineId, "pipelineId is null");
+        this.splitId = requireNonNull(splitId, "splitId is null");
+        this.operatorId = requireNonNull(operatorId, "operatorId is null");
+        this.enabled = isEnabled;
+    }
+
+    public static Tracer createBasicTracer(Consumer<TracerEvent> emitter, Function<Map<String, Object>, String> payloadEncoder, String nodeId, URI uri, String queryId, boolean enabled)
+    {
+        return (new Builder(emitter, payloadEncoder, nodeId, uri, queryId, enabled)).build();
```

Comment:
Parenthesis around `new Builder` are unnecessary. Consider using a factory method in the builder to improve readability:

```
builder(...).build()
```